examples/flutter_gallery/lib/demo/material/fab_motion_demo.dart

```diff
@@ -145,3 +156,71 @@ class _TopStartFloatingActionButtonLocation extends FloatingActionButtonLocation
     return new Offset(fabX, fabY);
   }
 }
+
+/// Provider of common logic for [FloatingActionButtonLocation]s that
+/// dock to the [BottomAppBar].
+abstract class _DockedFloatingActionButtonLocation extends FloatingActionButtonLocation {
```

Comment:
Could this be a mixin instead?  Not that it needs to be, but since it's just providing functionality and no state, it seems like a candidate for that.